CMSgov/bcda-app
# BCDA-1130 Auth: increase security of stored hashes

### Fixes [BCDA-1130](https://jira.cms.gov/browse/BCDA-1130)
We are following the recommendations in [NIST 800-63b](https://pages.nist.gov/800-63-3/sp800-63b.html#sec5) to decrease the likelihood of a successful attack by implementing the pbkdf2 key-derivation protocol when storing our secrets. This adds a random "salt" to the secret and drastically increases the time to brute-force an attack on the stored secrets by iterating the hash more than 10,000 times.

### Proposed changes:
- Import `golang.org/x/crypto/pbkdf2`
- Use the new library when creating/storing hashes
- Tune the number of iterations as high as possible
- Document and test to ensure we stay safe and compatible

### Data Migration Strategy
No changes are made to the database in this code.  However, it does break any existing client secrets. 
 At the moment client id's & secrets should only exist in the `dev` and `test` environments.  Except for a set of credentials used by Jenkins, they should be ephemeral.  The migration strategy is therefore:
- Delete any existing client ID's and secrets from the `acos` table
- Deploy this code
- Set up new credentials for Jenkins to use in smoke and postman tests in `dev` and `test`

### Security Implications
See SIA for details.  Note that we are comfortable with the idea that alterations to this code (including number of iterations) will force credential rotation in production.

### Testing
- Make sure to run `dep ensure` before running tests
- If you wish to see the actual hash execution time on your computer
   - Set the format for `gotestsum` in `/unit_test.sh`: Line 27: . . . `gotestsum --format standard-verbose` . . .
   - Search the output of `make unit-test` for `Hash creation`

### Feedback Requested
- Are there additional areas of code which could use comments?
- Do any existing comments need to be clarified?
- Are there additional unit, integration, smoke, or other tests that we should create?
- Should we adjust the initial number of hash iterations?  The security recommendation is to have them as high as we can tolerate.  This code aims to add 1s to all calls to `/auth/token`--can we tolerate more?
